packages/core/strapi/src/services/document-service/document-repository.ts

```diff
@@ -0,0 +1,88 @@
+import { Strapi, Common, Documents } from '@strapi/types';
+import createDocumentService from '.';
+import createMiddlewareManager from './middlewares';
+
+/**
+ * TODO:
+ *  - Transactions
+ *  - _executeMiddlewares
+ *
+ * Repository to :
+ * - Access documents via actions (findMany, findOne, create, update, delete, ...)
+ * - Execute middlewares on document actions
+ * - Apply default parameters to document actions
+ *
+ * @param strapi
+ * @param options.defaults - Default parameters to apply to all actions
+ * @param options.parent - Parent repository, used when creating a new repository with .with()
+ * @returns DocumentRepository
+ *
+ * @example Access documents
+ * const article = strapi.documents('api::article.article').create(params)
+ * const allArticles = strapi.documents('api::article.article').findMany(params)
+ *
+ */
+export const createDocumentRepository = (
+  strapi: Strapi,
+  { defaults = {} }: { defaults?: any } = {}
+): Documents.Repository => {
+  const documents = createDocumentService({ strapi, db: strapi.db! });
+  const middlewareManager = createMiddlewareManager();
+
+  function create<TContentTypeUID extends Common.UID.ContentType>(
+    uid: TContentTypeUID
+  ): Documents.RepositoryInstance<TContentTypeUID> {
+    return {
+      async findOne(id, params = {}) {
+        return middlewareManager.run(
+          {
+            action: 'findOne',
+            uid,
+            params,
+            options: { id },
+          },
+          ({ params }) => documents.findOne(uid, id, params)
+        );
+      },
```

Comment:
@alexandrebodin we talked about using a "compose" pattern to run middlewares.

I am not able to find something that makes it clear 🤔 
```
      findOne: compose(
        middlewareManager.getFor(uid, 'findOne'), 
        () =>documents.findOne(uid, id, params)
      )({
        ctx: {
          action: 'findOne',
          uid,
          params,
          options: { id },
        },
      }),
```

I abstracted away the middleware registration into `middlewareManager`, I don't know if there are other alternatives that could help here.

Do you have some examples of how we use the compose pattern in our repo? I only found one simple example with a hardcoded middleware.